juju/juju
# Restructure charms client

Create new LocalCharmClient, which steals the method AddLocalCharm from
the original charms client.

This is because in future we will want to add extra dependencies to
AddLocalCharm, which are not relevant to the rest of the api methods.

(We are soon to add an S3-compatible endpoint to the controller for
uploading local charms, meaning we will need to inject an S3 client into
the api so that this Put operation can be performed)

Only a small number of the charm.Clients we construct actually need to
be able to call AddLocalCharm. Only two in production code (once in deploy,
once in refresh)

All non-deploy/refresh code (e.g. diffbundle, ssh_container,
charmresources) will just keep the reduced client, meaning they never
need to worry about these extra dependencies soon to be introduced
## Checklist


## QA steps

Since this is quite a simple change, simply compiling should do. However, for completeness:

```
$ juju bootstrap lxd lxd
$ juju add-model m
$ juju deploy ubuntu
$ juju refresh ubuntu --path /path/to/charm

$ juju add-model m2
$ juju deploy juju-qa-bundle-test
$ juju diff-bundle juju-qa-bundle-test

$ juju add-model m3
$ juju deploy ~/charms/ubuntu

$ juju add-model m4
$ cat ./bundle.yaml
applications:
  ubuntu:
    charm: /gome/jack/charms/ubuntu
$ juju deploy ./bundle.yaml
```
